### ISSUE TITLE
Task spawning and invalid cycle points

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc-flow#4514]
```ini
P2 = "foo[-P1] => bar"  # (sequence P2, with erroneous P1 offset)
```
^ this fails validation in both 7 and 8 with *no cycling sequences defined for foo*

But this:
```ini
P2 = """
   foo
   foo[-P1] => bar
"""
```
^ at Cylc 7:
- passes validation (a cycling sequence is defined for `foo`)
- `foo.1` and `bar.1` run (pre-initial ignore)
- `foo` runs at every `P2` cycle ... 
- ... until suite stalls with prespawned `bar`s waiting on non-existent off-sequence `foo`s

^ at Cylc 8:
- passes validation
- `foo.1` and `bar.1` run (pre-initial ignore)
- `foo` runs at every `P2` cycle
- it does not stall, and no `bar`s run

At first glance this is the expected result. Cylc 8 doesn't pre-spawn waiting task so we should not expect a stall here.

But on closer inspection...

At Cylc 7 the graph above is interpreted as :
- `foo` is valid at every `P2` point
- `bar` at every `P2` point depends on `foo` at non-existent off-sequence points

At Cylc 8, the same is true for prerequisite definition, but for spawning purpose (spawn on demand) the second line gets interpreted as:
- `foo => bar[+P1]`

So `bar` doesn't run (at Cylc 8) because `foo` on `P2` points tries to spawn `bar` at off-sequence points, which is technically not correct.

I don't think that matters in practice, because:
- it works correctly when there isn't an offset error in the graph
- the result is the same (albeit for a bass-ackward reason) when there is an offset error: the child `bar` doesn't run

And `bar` prerequisites are still correct if they get spawned bother another task:
```
P2 = """
    foo
    foo[-P1] => bar
    qux => bar
"""
^ This correctly stalls at Cylc 8, with `bar` on `P2` points waiting on non-existent off-sequence `foo`.
 
Still, it raises some questions:
- can we detect these problems at validation?
- can we detect that `foo[-P1] => bar` is invalid if `foo` only exists at `P2` points, to avoid attempting to interpret it incorrectly as `foo => bar[+P1]`

**Comment by hjoliver** [comment]
Related example that doesn't have an error in the graph:

```ini
[scheduling]
    initial cycle point = next(T06)
    runahead limit = P1
    [[graph]]
        PT6H = "waz"
        T06 = "waz[-PT6H] => foo"
```
^ `waz` runs at every `PT6H` point, and `foo.T06` triggers off of `waz.T00` only.

In Cylc 8, for spawning purposes, we try to spawn  `waz => foo[+PT6H]` at every `PT6H` point which is technically incorrect, but it works because `foo` is invalid except at `T06`.

**Comment by TomekTrzeciak** [comment]
I've been aware of this behaviour for some time and always felt a bit uneasy about it. I think this matters, particularly for sequences with exclusions (say, you exclude every New Years day on a `P1D` sequence only to find out that some dependency has stalled your suite while you're on your Christmas break).

For the suite generating framework I've written for IMPROVER I avoided this problem by taking a different approach. I've made the cycling frequency into a property of the task itself rather than the graph of tasks. Then, for a dependency like `foo[P1D] => bar` the valid cycles are worked out automatically by shifting the `foo` sequence by `P1D` offset and taking intersection with the `bar` sequence. An empty sequence intersection indicates an impossible dependency and exception is thrown. Generally, I've found this easier to work with and reason about, in particular for large and complex graphs (only one place where the task cycling is defined to worry about).

**Comment by hjoliver** [comment]
> I've made the cycling frequency into a property of the task itself rather than the graph of tasks. 

Haha, the earliest versions of Cylc did exactly that, before we had a dependency graph to define prerequisites and cycling.

**Comment by hjoliver** [comment]
> I think this matters,...

Presumably if we can detect and fail this sort of graphing error at validation time, you'd be happy?

**Comment by TomekTrzeciak** [comment]
> > I've made the cycling frequency into a property of the task itself rather than the graph of tasks.
> 
> Haha, the earliest versions of Cylc did exactly that, before we had a dependency graph to define prerequisites and cycling.

Maybe this should be revisited 😉 

> > I think this matters,...
> 
> Presumably if we can detect and fail this sort of graphing error at validation time, you'd be happy?

I don't think it's a major problem in practice, but it would help if the validation could catch it.
